### ISSUE TITLE
Emit synchronous top-level export

### ISSUE THREAD

**Issue by WorldSEnder** [wasm-tool/rollup-plugin-rust#35]
`wasm-bindgen` emits, additionally to the `export default init`, an `export { initSync }` for synchronous initialization. Unfortunately, this export is not passed on with this plugin. See also microsoft/playwright#17221 where this could be useful.

I would be able and willing to write a PR for this, let me know what you think. I am aware that it will most likely require the wasm file to be included verbatim in the javascript, as I don't see another way to load the wasm bytes synchronously.

**Comment by Pauan** [comment]
`initSync` doesn't load the Wasm, it requires you to load the Wasm in some other way.

Browsers don't support synchronous loading of Wasm (except inside of Web Workers, and even that is iffy).

`initSync` is primarily intended to be used in multi-threaded Wasm, where you send the Wasm module to other Web Workers.

Instead, you can write your unit tests like this:

```ts
import load from '../Cargo.toml';

const wasm = load();

test('foo', async ({ page }) => {
  const { YewComponent } = await wasm;
  // Rest of unit test here
});

test('bar', async ({ page }) => {
  const { YewComponent } = await wasm;
  // Rest of unit test here
});
```

It will only load the Wasm one time, even if you use it in multiple tests. And after Microsoft adds in top-level `await` support, you'll be able to use the simpler version:

```ts
import wasm from '../Cargo.toml';
const { YewComponent } = await wasm();

test('foo', async ({ page }) => {
  // Rest of unit test here
});

test('bar', async ({ page }) => {
  // Rest of unit test here
});
```

**Comment by WorldSEnder** [comment]
> Browsers don't support synchronous loading of Wasm (except inside of Web Workers, and even that is iffy).

Yes I know, though the option of including the wasm is already in this lib, but apparently there's still no synchronous loading, unless I missed this. I would be honestly fine with this until test frameworks adapt.

> Instead, you can write your unit tests like this:

Sadly not, although this is test framework specific, see the linked bug report above. They have an additional brittle logic figuring out which components are being tested, and with the workarounds you posted below, their framework detects none of them.

Basically, the way it works on their side is that it looks specifically for import statements of the form: `import { ... } from '...';` where the component name is also mentioned in some call to `render(<SomeComp />` - so the workarounds don't mention that name and are disregarded. Pretty inconvenient, but I'll have to live with it. A shim that re-exports all components would still have to synchronously detect all the components' names, or let the user hard-code them.

The easiest fix, really, at least in the current state of affairs, seems to be to include the wasm as an `ArrayBuffer` and synchronously initialize :/

----

I would be fine with the synchronous export being gated behind `options.inlineWasm`.

**Comment by Pauan** [comment]
> Basically, the way it works on their side is that it looks specifically for import statements of the form: import { ... } from '...'; where the component name is also mentioned in some call

If you do something like this, does their detection work?

```ts
import type { YewComponent } from './shim';
import load from '../Cargo.toml';

const wasm = load();

test('foo', async ({ page }) => {
  const { YewComponent } = await wasm;
  // Rest of unit test here
});
```

In that case the shim would just need to export a type, like this:

```ts
export interface YewComponent {}
```

**Comment by WorldSEnder** [comment]
> If you do something like this, does their detection work?

With that method, for some reason, it doesn't use this rollup plugin to compile the code, and doesn't transform the
`import load from '../Cargo.toml';` import. Honestly, their auto-detection seems to do insane/mystical things. I think the point is that the file gets split into a part that gets transformed with the correct rollup plugins and one part that gets transformed with a smaller set of plugins, but I'm really not sure. I'd have to read their code in-depth to understand what's going on :sweat_smile: 

I will push the code at some point, if you really want to start debugging this yourself.

**Comment by Pauan** [comment]
Well, that sounds terrible. In the long run the plan is to remove the two-stage loading and just do this:

```ts
import { YewComponent } from '../Cargo.toml';
```

However, we can't do that until top-level await is supported in Rollup:

https://github.com/rollup/rollup/issues/3623

https://github.com/wasm-tool/rollup-plugin-rust/issues/31

So I'm curious, if you create a shim file like this, does it work in playwright?

```ts
import wasm from '../Cargo.toml';
const { YewComponent } = await wasm();
export { YewComponent };
```

And then import it as usual:

```ts
import { YewComponent } from './shim';

test('foo', async ({ page }) => {
  // Rest of unit test here
});
```

**Comment by WorldSEnder** [comment]
> So I'm curious, if you create a shim file like this, does it work in playwright?
> 
> ```ts
> import wasm from '../Cargo.toml';
> const { YewComponent } = await wasm();
> export { YewComponent };
> ```

I'd have to investigate further, but at least on the first try, it runs into the same problem of `./shim.{mjs,cjs,js}` not being allowed a top-level await. I'll see if maybe there is some plugin available that correctly transforms this and does run correctly. It sounds like a good idea.

**Comment by Pauan** [comment]
If the shim does work, then I would be okay with adding a setting that will generate top-level await (like described in [here](https://github.com/wasm-tool/rollup-plugin-rust/issues/31)). That should completely fix your issue.

However, if the top-level await doesn't work, then things get a lot trickier.

**Comment by Pauan** [comment]
I just released version 2.3.0. This adds two new options:

* `experimental.directExports` changes the way modules are created, so now you can do this:

   ```ts
   import { YewComponent } from '../Cargo.toml';
   ```

* `experimental.synchronous` causes the Wasm to be loaded synchronously. This only works with `inlineWasm: true`.

If you combine those two settings, it should give you what you originally asked for.

**Comment by Pauan** [comment]
@WorldSEnder How are you running Rollup with Playwright?
